src/modules/peek/Peek.FilePreviewer/Controls/ShellPreviewHandlerControl.xaml.cs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.UI;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Peek.FilePreviewer.Previewers;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.UI.Shell;
+using Windows.Win32.UI.WindowsAndMessaging;
+
+namespace Peek.FilePreviewer.Controls
+{
+    public unsafe sealed partial class ShellPreviewHandlerControl : UserControl
+    {
+        // Making this into a DependencyProperty causes a InvalidCastException
+        private IPreviewHandler? _source;
+
+        private HWND containerHwnd;
+        private WNDPROC containerWndProc;
+        private RECT controlRect;
+
+        public static readonly DependencyProperty LoadingStateProperty = DependencyProperty.Register(
+            nameof(LoadingState),
+            typeof(PreviewState),
+            typeof(ShellPreviewHandlerControl),
+            new PropertyMetadata(PreviewState.Uninitialized));
+
+        public PreviewState? LoadingState
+        {
+            get { return (PreviewState)GetValue(LoadingStateProperty); }
+            set { SetValue(LoadingStateProperty, value); }
+        }
+
+        public IPreviewHandler? Source
+        {
+            get => _source;
+            set
+            {
+                _source = value;
+                SourcePropertyChanged();
+            }
+        }
+
+        public ShellPreviewHandlerControl()
+        {
+            InitializeComponent();
+
+            containerWndProc = ContainerWndProc;
+        }
+
+        private void SourcePropertyChanged()
+        {
+            if (Source != null)
+            {
+                UpdatePreviewerTheme();
+
+                try
+                {
+                    // Attach the preview handler to the container window
+                    Source.SetWindow(containerHwnd, (RECT*)Unsafe.AsPointer(ref controlRect));
+                    Source.DoPreview();
+
+                    PInvoke.ShowWindow(containerHwnd, SHOW_WINDOW_CMD.SW_SHOW);
+                }
+                catch
+                {
+                }
+            }
+            else
+            {
+                PInvoke.ShowWindow(containerHwnd, SHOW_WINDOW_CMD.SW_HIDE);
+            }
+        }
+
+        private void UpdatePreviewerTheme()
+        {
+            if (Source is IPreviewHandlerVisuals visuals)
+            {
+                try
+                {
+                    switch (ActualTheme)
+                    {
+                        case ElementTheme.Light:
+                            visuals.SetBackgroundColor(new COLORREF(0x00f3f3f3));
```

Comment:
Where are these color values taken from?

Reply:
They're the Mica fallback colors.